File: AOC_2023/Day 23/23.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <algorithm>
#include <set>
#include <map>
#include <queue>

using namespace std;

struct edge{
    pair<int,int> vertex;
    int weight;
};

class Solution
{
    public:
    vector<string> grid;
    pair<int,int> start, goal;
    int n,m;
    pair<int,int> dirs[4] = {{0,1},{0,-1},{1,0},{-1,0}};
    map<char,int> slopetoDir = {{'>', 0}, {'<', 1}, {'v', 2}, {'^', 3}};
    void init(ifstream& input){
        string line;
        while (getline(input, line)) grid.push_back(line);
        n = grid.size();
        m = grid[0].size();
        for (int j = 0; j < m; j++){
            if(grid[0][j] == '.'){
                start = {0,j};
                break;
            }
        }
        for (int j = 0; j < m; j++){
            if(grid[n-1][j] == '.'){
                goal = {n-1,j};
                break;
            }
        }
    }

    bool inBounds(pair<int,int> p){
        return p.first >= 0 && p.first < n && p.second >= 0 && p.second < m;
    }

    int helper(bool part2){
        // find all vertices that are paths(.) that can be junctions and add them to the graph
        set<pair<int,int>> vertices;
        vertices.insert(start);
        vertices.insert(goal);
        for(int i = 0; i < n; i++){
            for(int j = 0; j < m; j++){
                int neighbors = 0;
                // check if the current cell is a junction i.e., having more than 2 paths to go 
                // else it is just a normal straight path or a dead end
                for(auto& dir : dirs){
                    if(inBounds({i+dir.first,j+dir.second}) && grid[i+dir.first][j+dir.second] != '#') neighbors++;
                }
                if(neighbors > 2 && grid[i][j] != '#') vertices.insert({i,j});
            }
        }
        // build the graph with the vertices and edges
        map<pair<int,int>,vector<edge>> graph;
        for(auto& v : vertices){
            queue<edge> q;
            // bfs to find all the paths from the current vertex
            q.push({v,0});
            set<pair<int,int>> visited;
            while(!q.empty()){
                auto curr = q.front();
                q.pop();
                // if the current vertex is already visited then skip it
                if(visited.count(curr.vertex)) continue;
                visited.insert(curr.vertex);
                // if the current vertex is a junction then add it to the graph
                if(vertices.count(curr.vertex) && curr.vertex != v){
                    graph[v].push_back(curr);
                    continue;
                }
                // else add all the neighbors of the current vertex to the queue
                for(auto& dir : dirs){
                    auto next = make_pair(curr.vertex.first+dir.first,curr.vertex.second+dir.second);
                    if(inBounds(next) && grid[next.first][next.second] != '#'){
                        if(!part2 && slopetoDir.count(grid[next.first][next.second]) && dirs[slopetoDir[grid[next.first][next.second]]] != dir) continue;
                        q.push({next,curr.weight+1});
                    }
                }
            }
        }
        // dfs to find the longest path from start to goal
        int result = 0;
        // visited set to avoid cycles
        set<pair<int,int>> visited;
        // dfs function
        auto dfs = [&](auto&& dfs, pair<int,int> curr, int weight, set<pair<int,int>>& visited){
            // if the current vertex is already visited then skip it
            if(visited.count(curr)) return;
            // else add it to the visited set
            visited.insert(curr);
            // if the current vertex is the goal then update the result
            if(curr == goal){
                result = max(result,weight);
                visited.erase(curr);
                return;
            }
            // visit all the neighbors of the current vertex
            for(auto& e : graph[curr]) dfs(dfs,e.vertex,weight+e.weight,visited);
            // remove the current vertex from the visited set
            visited.erase(curr);
        };
        // call the dfs function
        dfs(dfs,start,0,visited);
        return result;
    }

    void part1(){
        cout << "Part 1: " << helper(false) << "\n";
    }

    void part2(){
        cout << "Part 2: " << helper(true) << "\n";
    }

};

int main(){
    ifstream in("input.txt");
    if (!in.is_open()){
        cerr << "Could not open the file - '" << endl;
        return EXIT_FAILURE;
    }
    Solution s;
    s.init(in);
    s.part1();
    s.part2();
    return 0;
}

File: AOC_2023/Day 23/23.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

type edge struct {
	vertex pair
	weight int
}

type pair struct {
	first, second int
}

type Solution struct {
	grid        []string
	start, goal pair
	n, m        int
	dirs        [4]pair
	slopetoDir  map[byte]int
}

func (s *Solution) init(input *os.File) error {
	s.grid = make([]string, 0)
	s.dirs = [4]pair{{0, 1}, {0, -1}, {1, 0}, {-1, 0}}
	s.slopetoDir = map[byte]int{'>' : 0, '<' : 1, 'v' : 2, '^' : 3}
	scanner := bufio.NewScanner(input)
	for scanner.Scan() {
		s.grid = append(s.grid, scanner.Text())
	}
	if err := scanner.Err(); err != nil {
		return err
	}
	s.n = len(s.grid)
	s.m = len(s.grid[0])
	for j := 0; j < s.m; j++ {
		if s.grid[0][j] == '.' {
			s.start = pair{0, j}
			break
		}
	}
	for j := 0; j < s.m; j++ {
		if s.grid[s.n-1][j] == '.' {
			s.goal = pair{s.n - 1, j}
			break
		}
	}
	return nil
}

func (s *Solution) inBounds(p pair) bool {
	return p.first >= 0 && p.first < s.n && p.second >= 0 && p.second < s.m
}

func (s *Solution) helper(part2 bool) int {
	vertices := make(map[pair]bool)
	vertices[s.start] = true
	vertices[s.goal] = true
	for i := 0; i < s.n; i++ {
		for j := 0; j < s.m; j++ {
			neighbors := 0
			for _, dir := range s.dirs {
				if s.inBounds(pair{i + dir.first, j + dir.second}) && s.grid[i+dir.first][j+dir.second] != '#' {
					neighbors++
				}
			}
			if neighbors > 2 && s.grid[i][j] != '#' {
				vertices[pair{i, j}] = true
			}
		}
	}
	graph := make(map[pair][]edge)
	for v := range vertices {
		q := []edge{{v, 0}}
		visited := make(map[pair]bool)
		for len(q) > 0 {
			curr := q[0]
			q = q[1:]
			if visited[curr.vertex] {
				continue
			}
			visited[curr.vertex] = true
			if vertices[curr.vertex] && curr.vertex != v {
				graph[v] = append(graph[v], curr)
				continue
			}
			for _, dir := range s.dirs {
				next := pair{curr.vertex.first + dir.first, curr.vertex.second + dir.second}
				if s.inBounds(next) && s.grid[next.first][next.second] != '#' {
					val, ok := s.slopetoDir[s.grid[next.first][next.second]]
					if !part2 && ok && s.dirs[val] != dir {
						continue
					}
					q = append(q, edge{next, curr.weight + 1})
				}
			}
		}
	}
	result := 0
	visited := make(map[pair]bool)
	var dfs func(pair, int, map[pair]bool)
	dfs = func(curr pair, weight int, visited map[pair]bool) {
		if visited[curr] {
			return
		}
		visited[curr] = true
		if curr == s.goal {
			result = max(result, weight)
		}
		for _, e := range graph[curr] {
			dfs(e.vertex, weight+e.weight, visited)
		}
		delete(visited, curr)
	}
	dfs(s.start, 0, visited)
	return result
}

func (s *Solution) part1() {
	fmt.Println("Part 1:",s.helper(false))
}

func (s *Solution) part2() {
	fmt.Println("Part 2:",s.helper(true))
}

func main() {
	file, err := os.Open("input.txt")
	if err != nil {
		fmt.Println("Error reading file:", err)
		return
	}
	defer file.Close()
	var s Solution
	err = s.init(file)
	if err != nil {
		fmt.Println("Error initializing solution:", err)
		return
	}
	s.part1()
	s.part2()
}